### ISSUE TITLE
ERROR spoon.support.compiler.jdt.JDTCommentBuilder

### ISSUE THREAD

**Issue by SJdatageek** [INRIA/spoon#4162]
[main] ERROR spoon.support.compiler.jdt.JDTCommentBuilder - "/* example comment */" commented line cannot be added into the AST, with parent class spoon.support.reflect.code.CtBinaryOperatorImpl

**Comment by slarse** [comment]
Hi @SJdatageek,

Could you provide some more information here about how you encounter this error? What project are you trying to analyze/transform, and what does your Spoon code look like?

**Comment by SJdatageek** [comment]
Hi, I'm facing this error when there's a comment inside if statement in the code.. ideally this should throw a warning. Can't give anymore details on the project and code.

**Comment by slarse** [comment]
> I'm facing this error when there's a comment inside if statement in the code

Is it in the condition or the then/else branches of the if statement?

> ideally this should throw a warning

How do you mean? Ideally, there shouldn't be a problem adding the comment, right?

> Can't give anymore details on the project and code.

I unfortunately can't provide any help without more information, I have no idea of what's causing this. Can you provide scrubbed version of the if-statement causing the problem? Just remove any sensitive information, rename all variables, etc, but ensure that the error still occurs.

Also, what version of Spoon are you using, with what settings (compliance level etc)?

**Comment by slarse** [comment]
@monperrus Have you seen this error before? I don't even know where it's originating from, I can't see anything in `JDTCommentBuilder` that would throw an exception like that, or write to the log file with anything remotely resembling it.

**Comment by MartinWitt** [comment]
I believe he failed this check here [https://github.com/INRIA/spoon/blob/2fff4c50266f79bc06b961e3b01ca6c6e0fcc1ab/src/main/java/spoon/support/compiler/jdt/JDTCommentBuilder.java#L502](https://github.com/INRIA/spoon/blob/2fff4c50266f79bc06b961e3b01ca6c6e0fcc1ab/src/main/java/spoon/support/compiler/jdt/JDTCommentBuilder.java#L502), but as you said without any information there is nothing to do.

Have you thought about adding an issue template, which has fields for the always needed information like spoon version etc.?

**Comment by SJdatageek** [comment]
> Is it in the condition or the then/else branches of the if statement?
Ex: if(((a>100 && a<0)/*this is a comment*/) || (a==b)) { } else {}
> Also, what version of Spoon are you using, with what settings (compliance level etc)?
Version- 9.1.0-beta-10

**Comment by slarse** [comment]
Thanks @MartinWitt that's the stuff. I don't understand how my text search did not turn that up, I just searched for "cannot" inside of JDTCommentBuilder.

@SJdatageek I tried parsing the following class and can't reproduce the logging message.

```java
class Test  {
    void test() {
        if(((a>100 && a<0)/*this is a comment*/) || (a==b)) { } else {}
    }
}
```
I also tried parsing it as a snippet.

```java
CtStatement statement = launcher.getFactory()
		.createCodeSnippetStatement("if(((1>100 && 1<0)/*this is a comment*/) || (1==2)) { } else {}")
		.compile();
```

But I can't reproduce the problem.

It's correct to log this as an error, because if a comment has no parent after being inserted it's in an invalid state (or it wasn't inserted).

Without a complete example that reproduces the problem, we unfortunately can't help.

**Comment by SJdatageek** [comment]
@slarse can you please try with the below example once more and see if it reproduces the error for you?
`class Test {
public boolean testA() throws Exception {
		String str1 = "this is an example string";
		/* some comment here */
		// another example comment
		// one more example comment
		// comment.
		

		// code fix for ends Here

		// more comments
		if (((str1 != null && str1.length > 0) 
				/* this is the problem comment */		
				)
				|| (str1.equals("this is an example string"))) {
			return true;
		} else {
			return false;
		}
	}
}`

**Comment by slarse** [comment]
@SJdatageek Yes, that works!

Thanks, now we have something to work with :)

**Comment by slarse** [comment]
I have created a minimal class to reproduce the problem:

```java
class Test  {
    void test() {
        if((true
                /*this is a comment*/)) { }
    }
}
```

In the case of the if condition, the problem appears to be when a comment is inside parentheses, and there's a line break before the comment. However, I also encounter the same problem with something like this:

```java
int a = (1 /*test*/);
```

But it doesn't occur without the parentheses:

```java
int a = 1 /*test*/;
```

Soo... Something with parentheses and potentially newlines. I'll have to get back to this later, but it's a start!

**Comment by SJdatageek** [comment]
Yes, I'm getting this error when as you mentioned above,  a comment is inside parentheses, and there's a line break before the comment.

**Comment by I-Al-Istannen** [comment]
I thought that was just not implemented and nobody cares because that error pops up a lot. But here are two more:
```
19:33:11.207 [main] ERROR spoon.support.compiler.jdt.JDTCommentBuilder - "/* ... */" cannot be added into the AST, with parent class spoon.support.reflect.code.CtVariableReadImpl at (/home/i_al_istannen/Programming/Discord/JavadocApi/java.base/java/lang/invoke/LambdaFormEditor.java:686), please report the bug by posting on https://github.com/INRIA/spoon/issues/2482


19:33:18.886 [main] ERROR spoon.support.compiler.jdt.JDTCommentBuilder - "// || (Axis.PRECEDINGSIBLING == axis)" cannot be added into the AST, with parent class spoon.support.reflect.code.CtBinaryOperatorImpl at (/home/i_al_istannen/Programming/Discord/JavadocApi/java.xml/com/sun/org/apache/xpath/internal/axes/WalkerFactory.java:788), please report the bug by posting on https://github.com/INRIA/spoon/issues/2482
```

Code in question is in JDK 16, but the excerpt is:
```java
        Name callCombiner = new Name(new NamedFunction(arrayCollector, Intrinsic.NEW_ARRAY),
                                        (Object[]) /*...*/ newParams);
```
in LambdaFormEditor.java and
```java
  public static boolean isDownwardAxisOfMany(int axis)
  {
    return ((Axis.DESCENDANTORSELF == axis) ||
          (Axis.DESCENDANT == axis)
          || (Axis.FOLLOWING == axis)
//          || (Axis.FOLLOWINGSIBLING == axis)
          || (Axis.PRECEDING == axis)
//          || (Axis.PRECEDINGSIBLING == axis)
          );
  }
```
in the WalkerFactory.

**Comment by slarse** [comment]
@I-Al-Istannen Yep that's the same bug. It seems to be all about parenthesized expressions. I hope to have a look during the weekend.

**Comment by jose** [comment]
Is there any news on this?  It seems this issue is still present in version 10.2.0-beta-11.

**Comment by slarse** [comment]
To my knowledge no one has worked on this yet.
